msteams-platform/concepts/bots/bots-files.md

```diff
@@ -0,0 +1,152 @@
+---
+title: Sending and receiving files from a bot
+description: Describes how to send and receive files from a bot
+keywords: teams bots files send receive
+ms.date: 04/17/2018
+---
+# Send and receive files through your bot
+
+>**Note:** This feature is currently available in Developer Preview.
+
+Your bot can send and receive files with users in the personal (1:1) context. You can use this to implement expense reporting, image recognition, file archival, e-signatures, and other scenarios involving direct manipulation of file content. Files shared in Teams typically appear as cards, and allow rich in-app viewing. In channels, you can also post messages with card attachments referencing existing SharePoint files.
+
+## Overview
+
+One way of doing this is using the Microsoft Graph APIs for OneDrive and SharePoint, [as documented here](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/). This method requires obtaining ongoing access to the user's OneDrive folder through standard OAuth2 authorization flow. This documentation describes a simple alternative mechanism if you only need to send file content as a result of direct user interaction, e.g. sending a message. This API is provided as part of the Microsoft Teams Bot Platform.
+
+## Receiving files in 1:1 chat
+
+When a user sends a file to your bot, the file is first uploaded to the user's OneDrive for Business storage. Your bot will then receive a message activity notifying you of the user upload. The activity will contain file metadata, such as its name and the content URL, and the content URL's expiry time (e.g. 24 hours). You can directly read from this URL to fetch its binary content.
+
+### Message activity with file attachment example
+
+```json
+{
+  "attachments": [{
+    "contentType": "application/vnd.microsoft.teams.file.download.info",
+    "contentUrl": "https://contoso.sharepoint.com/personal/johnadams_contoso_com/Documents/Applications/file_example.txt",
+    "name": "file_example.txt",
+    "content": {
+      "downloadUrl" : "https://download.link",
+      "uniqueId": "1150D938-8870-4044-9F2C-5BBDEBA70C9D",
+      "fileType": "txt",
+      "etag": "123"
+    }
+  }]
+}
+```
+
+As a best practice, you should acknowledge the file upload by sending back a message to the user.
+
+## Uploading files to 1:1 chat
+
+Uploading a file to a user involves the following steps:
+
+1. Send a message to the user requesting permission to write the file. This message must contain a `FileConsentCard` attachment with the name of the file to be uploaded.
+2. If the user accepts the file download, your bot will receive an Invoke activity with a location URL.
+3. To transfer the file, your bot performs an HTTP POST directly into the provided location URL.
+4. Optionally, you can perform a message update on the original message, removing the consent card if you do not want to allow the user to accept further uploads of the same file.
```

Comment:
Larry, have you tried this? It doesn't work as you expect-- after the update, you're left with a broken consent card (no content, just Accept/Decline buttons). I think it's because message type changes, and we don't handle that properly. You have to switch away from the chat and back again before it renders correctly.